base/range.jl

```diff
@@ -355,20 +364,21 @@ let smallint = (Int === Int64 ?
         div(Int(r.stop)+Int(r.step) - Int(r.start), Int(r.step))
     end
 
-    length{T <: smallint}(r::UnitRange{T}) = Int(r.stop) - Int(r.start) + 1
+    length{T <: smallint}(r::AbstractUnitRange{T}) = Int(last(r)) - Int(first(r)) + 1
+    length{T <: smallint}(r::OneTo{T}) = Int(r.stop)
 end
 
 first{T}(r::OrdinalRange{T}) = convert(T, r.start)
+first{T}(r::OneTo{T}) = one(T)
 first{T}(r::FloatRange{T}) = convert(T, r.start/r.divisor)
 first{T}(r::LinSpace{T}) = convert(T, (r.len-1)*r.start/r.divisor)
 
-last{T}(r::StepRange{T}) = r.stop
-last(r::UnitRange) = r.stop
+last(r::OrdinalRange) = r.stop
```

Comment:
Not guaranteed, but I don't think there's a better fallback definition. (It also happens to work for all `OrdinalRange`s currently defined in `Base`.) Anyone can of course specialize